### PR TITLE
HPCC-21414 Add new Spark Thor XSD

### PR DESCRIPTION
**Pull request by kenrowland**
Add new XSD

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by kenrowland** [comment]
@rpastrana Please review

File: initfiles/componentfiles/configschema/xsd/sparkthor.xsd [hunk 1]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (rpastrana) [inline initfiles/componentfiles/configschema/xsd/sparkthor.xsd]: copyright header?

File: initfiles/componentfiles/configschema/xsd/sparkthor.xsd [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+<xs:schema
+        xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" attributeFormDefault="unqualified"
+        xmlns:hpcc="someuri">
+    <xs:include schemaLocation="types.xsd"/>
+    <hpcc:insert hpcc:schemaPath="/Environment/Software">
+        <xs:element name="SparkThorProcess" hpcc:itemType="sparkthor" hpcc:displayName="Spark Thor" maxOccurs="1" hpcc:class="component"
```
Comment (rpastrana) [inline initfiles/componentfiles/configschema/xsd/sparkthor.xsd]: I'm not sure how it appears in the current xsd schema, but the display name should be "SparkThor" or "Spark-Thor"

File: initfiles/componentfiles/configschema/xsd/sparkthor.xsd [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+                    hpcc:docid="Th.t1">
+            <xs:complexType>
+
+                <xs:sequence>
+                    <xs:element name="Instance" maxOccurs="unbounded" hpcc:class="elementSet"
```
Comment (rpastrana) [inline initfiles/componentfiles/configschema/xsd/sparkthor.xsd]: The new schema constructs might allow us to improve this portion of the spark-thor configuration...
I believe the intent here is to allow the admin to specify a cluster node as the spark master. By default, it should be tied to the ECLWatch ESP instance (if one exists)

File: initfiles/componentfiles/configschema/xsd/sparkthor.xsd [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+                                          hpcc:sourceKey="computer_netaddress"
+                                          hpcc:mirrorFrom="/Environment/Hardware/Computer[@netAddress]"/>
+                            <xs:attribute name="directory" type="xs:string"
+                                          hpcc:presetValue="c$\dafilesrv" hpcc:hidden="true"/>
+                        </xs:complexType>
+                    </xs:element>
+                </xs:sequence>
+
+                <xs:attributeGroup ref="buildInfo"/>
+                <xs:attribute name="name" type="xs:string" use="required" hpcc:tooltip="Name for this process"/>
```
Comment (rpastrana) [inline initfiles/componentfiles/configschema/xsd/sparkthor.xsd]: tooltip= "Name for the SparkThor instance"

File: initfiles/componentfiles/configschema/xsd/sparkthor.xsd [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+                <xs:attribute name="ThorClusterName" type="xs:string" hpcc:displayName="Thor Cluster Name" use="required"
```
Comment (rpastrana) [inline initfiles/componentfiles/configschema/xsd/sparkthor.xsd]: toolitp: "Thor cluster for spark workers to attach to"

**Review by rpastrana — CHANGES_REQUESTED** [review]
@kenrowland looks good. A few comments.

File: initfiles/componentfiles/configschema/xsd/sparkthor.xsd (hunk 5 above)
Comment (kenrowland) [reply initfiles/componentfiles/configschema/xsd/sparkthor.xsd]: Updated, New tip may be a bit pedantic since I removed the ending preposition.

**Review by kenrowland — COMMENTED** [review]
@rpastrana Please see updates from your comments.

File: initfiles/componentfiles/configschema/xsd/sparkthor.xsd (hunk 3 above)
Comment (rpastrana) [reply initfiles/componentfiles/configschema/xsd/sparkthor.xsd]: Let's change maxoccurs to 1, and use standard HW instance ref, remove dir attribute, not needed.

Comment (kenrowland) [reply initfiles/componentfiles/configschema/xsd/sparkthor.xsd]: Changes made.

**Review by kenrowland — COMMENTED** [review]
@rpastrana Please see latest changes.

File: initfiles/componentfiles/configschema/xsd/sparkthor.xsd [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" attributeFormDefault="unqualified"
+        xmlns:hpcc="someuri">
+    <xs:include schemaLocation="types.xsd"/>
+    <hpcc:insert hpcc:schemaPath="/Environment/Software">
+        <xs:element name="SparkThorProcess" hpcc:itemType="sparkthor" hpcc:displayName="Spark-Thor" minOccurs="0"
+                    maxOccurs="1" hpcc:class="component" hpcc:docid="Th.t1">
+            <xs:complexType>
+                <xs:sequence>
+                    <xs:element name="Instance" maxOccurs="1" hpcc:class="elementSet"
+                                hpcc:displayName="Bound Computers">
```
Comment (rpastrana) [inline initfiles/componentfiles/configschema/xsd/sparkthor.xsd]: display name = "Master Spark Node"

**Review by rpastrana — APPROVED** [review]
@kenrowland just one very minor change

**Comment by kenrowland** [comment]
@richardkchapman Please merge

**Pull request merged**